cli/proc_state.rs

```diff
@@ -235,17 +236,42 @@ impl ProcState {
       cli_options.cache_setting(),
       progress_bar.clone(),
     );
-    let npm_resolver = NpmPackageResolver::new(
-      npm_cache.clone(),
-      api,
-      cli_options.unstable()
-        // don't do the unstable error when in the lsp
-        || matches!(cli_options.sub_command(), DenoSubcommand::Lsp),
-      cli_options.no_npm(),
-      cli_options
-        .resolve_local_node_modules_folder()
-        .with_context(|| "Resolving local node_modules folder.")?,
-    );
+    let maybe_lockfile = if let Some(lockfile) = &lockfile {
+      if lockfile.lock().write {
+        None
+      } else {
+        Some(lockfile.clone())
+      }
+    } else {
+      None
+    };
+    let npm_resolver = if let Some(lockfile) = maybe_lockfile.clone() {
+      NpmPackageResolver::regenerate_from_lockfile(
+        npm_cache.clone(),
+        api,
+        cli_options.unstable()
+          // don't do the unstable error when in the lsp
+          || matches!(cli_options.sub_command(), DenoSubcommand::Lsp),
+        cli_options.no_npm(),
+        cli_options
+          .resolve_local_node_modules_folder()
+          .with_context(|| "Resolving local node_modules folder.")?,
+        lockfile,
+      )
+      .await?
+    } else {
+      NpmPackageResolver::new(
+        npm_cache.clone(),
+        api,
+        cli_options.unstable()
+          // don't do the unstable error when in the lsp
+          || matches!(cli_options.sub_command(), DenoSubcommand::Lsp),
+        cli_options.no_npm(),
+        cli_options
+          .resolve_local_node_modules_folder()
+          .with_context(|| "Resolving local node_modules folder.")?,
+      )
```

Comment:
There's lots of duplicate code here with the above. Could we consolidate it somehow? For example, in the lsp it seems like we'll need to be able to add a lockfile after the fact most likely in the `initialize` function so I think it's worth taking the time to future proof this for that... so probably we'll need an API like:

```rs
let npm_resolver = ...
npm_resolver.add_lockfile(lockfile).await?;
```